### PR TITLE
Optimiser: Fancy select for spend tiers.

### PR DESCRIPTION
**Pull request by ryan-rushton**
This just adds those last changes to my existing PR.

Screen Shot 2021-06-09 at 7 38 23 pm

Screen Shot 2021-06-09 at 7 38 11 pm

Screen Shot 2021-06-09 at 7 37 53 pm

### CODE REVIEW

File: src/app/dim-ui/Select.tsx [hunk 1]
```diff
@@ -91,29 +136,31 @@ export default function Select<T>({
...
-                    onClick: (e: any) => {
-                      e.nativeEvent.preventDownshiftDefault = item.disabled;
-                    },
-                  })}
-                >
-                  {item.content}
-                </div>
-              )
-          )}
+        <div style={dropdownStyle}>
```
Comment (ryan-rushton) [inline src/app/dim-ui/Select.tsx:139]: I didn't want to mess with how popper works so I just added another div and put the styles on that.

File: src/app/loadout-builder/filter/FilterBuilds.tsx [hunk 2]
```diff
@@ -37,6 +42,85 @@ export default function FilterBuilds({
   const workingStatRanges =
     statRanges || _.mapValues(statHashes, () => ({ min: 0, max: 10, ignored: false }));
 
+  const legendaryShardDef = defs?.InventoryItem.get(UpgradeMaterialHashes.legendaryShard);
+  const enhancementPrismDef = defs?.InventoryItem.get(UpgradeMaterialHashes.enhancementPrism);
+  const ascendantShardDef = defs?.InventoryItem.get(UpgradeMaterialHashes.ascendantShard);
+  const legendaryShardIcon = legendaryShardDef?.displayProperties.icon;
+  const enhancementPrismIcon = enhancementPrismDef?.displayProperties.icon;
+  const ascendantShardIcon = ascendantShardDef?.displayProperties.icon;
+  const ascendantShardName = ascendantShardDef?.displayProperties.name || 'Ascendant Shard';
```
Comment (ryan-rushton) [inline src/app/loadout-builder/filter/FilterBuilds.tsx:51]: Just using an english string as a fallback for these names but I never actually expect it to get used.

**Comment by bhollis** [comment]
Haven't looked at the code, but can we keep the elements left-justified?

**Comment by ryan-rushton** [comment]
Check that out now. I have accounted for icons or no icons.

File: src/app/dim-ui/Select.tsx [hunk 3]
```diff
@@ -70,10 +86,39 @@ export default function Select<T>({
...
+    };
+  }
+
+  if (maxDropdownWidth === 'button' && dropdownWidth === undefined && buttonRef.current) {
+    // Minus 2 because the menu has a thicker outline than the button border (2px vs 1px)
+    const width =
+      maxButtonWidth !== undefined
+        ? maxButtonWidth
+        : buttonRef.current.getBoundingClientRect().width - 2;
+    setDropdownWidth(width);
```
Comment (bhollis) [inline src/app/dim-ui/Select.tsx:104]: Don't do this in the render. You can use an effect for this.

File: src/app/loadout-builder/filter/FilterBuilds.tsx [hunk 4]
```diff
@@ -37,6 +42,93 @@ export default function FilterBuilds({
...
+      },
+    ],
+    [
+      ascendantShardIcon,
+      ascendantShardName,
+      enhancementPrismDef?.displayProperties.name,
+      enhancementPrismIcon,
+      legendaryShardDef?.displayProperties.name,
+      legendaryShardIcon,
+    ]
```
Comment (bhollis) [inline src/app/loadout-builder/filter/FilterBuilds.tsx:129]: Why not have this just depend on `defs` and do the lookups inside the `useMemo`?

Comment (ryan-rushton) [reply src/app/loadout-builder/filter/FilterBuilds.tsx]: Yeah I don't know why I didn't. It's done now.

File: src/app/dim-ui/Select.tsx (hunk 3 above)
Comment (ryan-rushton) [reply src/app/dim-ui/Select.tsx]: Done

**Pull request merged**